reports-scheduler/src/main/java/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/handler/ReportsJobActionHandler.java

```diff
@@ -0,0 +1,192 @@
+import org.elasticsearch.rest.RestResponse;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.SearchHits;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+
+import com.amazon.opendistroforelasticsearch.jobscheduler.spi.JobExecutionContext;
+import com.amazon.opendistroforelasticsearch.jobscheduler.spi.utils.LockService;
+import com.amazon.opendistroforelasticsearch.reportsscheduler.job.parameter.JobParameter;
+
+public class ReportsJobActionHandler extends AbstractActionHandler {
+  private final LockService lockService;
+  private static final Logger log = LogManager.getLogger(ReportsJobActionHandler.class);
+
+  /**
+   * Constructor function.
+   *
+   * @param client ES node client that executes actions on the local node
+   * @param channel ES channel used to construct bytes / builder based outputs, and send responses
+   * @param clusterService ES cluster service
+   */
+  public ReportsJobActionHandler(
+      NodeClient client, RestChannel channel, ClusterService clusterService) {
+    super(client, channel);
+    this.lockService = new LockService(client, clusterService);
+  }
+
+  public void getJob() {
+    // get all jobs from job_queue index
+    final SearchRequest searchRequest = new SearchRequest(JOB_QUEUE_INDEX_NAME);
+    final SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+    searchSourceBuilder.query(QueryBuilders.matchAllQuery());
+
+    //    // sort the document by enqueue_time
+    //    searchSourceBuilder.sort(new FieldSortBuilder("enqueue_time").order(SortOrder.DESC));
+    searchRequest.source(searchSourceBuilder);
+    client.search(
+        searchRequest,
+        ActionListener.wrap(
+            response -> onSearchJobResponse(response), exception -> onFailure(exception)));
+  }
+
+  private void onSearchJobResponse(SearchResponse response) {
+    final SearchHits hits = response.getHits();
+    final SearchHit[] searchHits = hits.getHits();
+    final List<SearchHit> searchHitsList = Arrays.asList(searchHits);
+    // randomize the jobs, for possible faster job retrieval
+    Randomness.shuffle(searchHitsList);
+    final Queue<SearchHit> searchHitsQueue = new LinkedList<>(searchHitsList);
+
+    log.info("original queue size: " + searchHitsQueue.size());
+
+    findFirstAvailableJob(searchHitsQueue);
+  }
+
+  /**
+   * Attempt to find single jobs that is not being locked, and send the job information back to
+   * reporting core for job execution.
+   *
+   * @param searchHitsQueue a queue which saves all jobs from .reports_scheduler_job_queue index
+   */
+  private void findFirstAvailableJob(Queue<SearchHit> searchHitsQueue) {
+    final SearchHit hit = searchHitsQueue.poll();
+    log.info("queue size after poll: " + searchHitsQueue.size());
+
+    if (hit != null) {
+      String jobId = hit.getId();
+      // set up jobParamater and jobContext that is required by lockService
```

Comment:
is this TODO: ?

Reply:
No, this is the comment for the following lines, which creates a "fake" jobparamater and jobcontext instance, required by initializing lockService instance